### DISCUSSION TITLE
How do I apply forces to bodies in a model?

### DISCUSSION THREAD

**Discussion by adamkewley** [ComputationalBiomechanicsLab/opensim-creator#864] [Q&A] [answered]
> AK: this question was ported from an email chain. It has been edited, redacted, etc. to fit the public discussions format.

I am working on OpenSim and OpenSim Creator I have several questions related to applying forces to things in the model:

- I have a whole human model and I would like to apply an external force on the shoulder. Should I apply it on the muscle or the bone?
- I have tried to use `PrescribedForce` but I don't see where to put the numerical data of the force, do you know if I should use another type of force?
- Once I have applied my force on my model, can I export my file in "xml file" and import it on opensim in order to use it in "inverse dynamics"?

**Comment by adamkewley** [comment, marked as answer]
I haven't made too many OpenSim models (I'm more of a developer) but I'll try to answer your questions :)

- Should I apply [the force] to the muscle or the bone?:
  - Forces are usually attached to physical frames in the model.
  - A body is a physical frame, but there are other types of physical frames (e.g. offset frames) that may be attached (in)directly to bodies.
  - Bodies are components in the model that are ultimately "affected" by the force. Therefore, any forces you add into a model must be attached (again, indirectly or directly) to a body frame.
  - A bone is usually a mesh file, not a body, but it may contain an offset frame that attaches to the body
  - A muscle is a force, not a body or a physical frame.
  - Therefore, attaching a force to a muscle--i.e. a force to a force--doesn't make sense and you shouldn't do it
  - Attaching a force to a body is probably what you want, and that body is probably the "parent" of the bone mesh (check the "Navigator" panel)
- I have used "prescribeforce" but I don't see where to put the numerical data of the force?:
  - PrescribedForce doesn't take straightforward numerical inputs that OpenSim Creator can expose to you via a simple UI. It takes a set of functions that describe positions in space.
  - The source code for it is here: https://github.com/opensim-org/opensim-core/blob/4.4/OpenSim/Simulation/Model/PrescribedForce.h
  - (the utility of the source code is that it explains every parameter better than OpenSim Creator can)
  - You must edit the XML manually to fill out a PrescribedForce (OpenSim Creator cannot help you :()
- Do you know if I should use another type of force?:
  - That depends entirely on what you are trying to model. There might be another type of force you can use, but I don't know what you are modelling
- Once I have applied my force on my model, can I export my file in "xml file" and import it on opensim in order to use it in "inverse dynamics"?
  - Yes. Saving whatever you create in the OpenSim Creator editor will produce an ".osim" (XML) model file that can be opened in OpenSim GUI and used in an IK analysis

**Reply by adamkewley** [reply]
> AK: reply from user in the email chain.

I have tried every type of "add/force" in OpenSim Creator and I don't have any feedback in my outputs or simulation.

So if I describe my model, maybe you can help me. So I have a full body. In my "navigator/bodyset/" I have all my bone which are separated in a bodyset. My goal is to add a force, for example 10N on my left scapula. For example, I tried an `ExternalForce`, but I don't understand what really is expected in each property field:

- Apply to a body : I think it's the bone where I want to apply my force so left scapula.
- Force expressed in body & point expressed in body: I don't know if I have to let ground.
- Point identifier & force identifier & torque identifier, they expected an identifer string localisation, but I don't know what it means
- The data source name , is it the name of my xml file of my body?

**Reply by adamkewley** [reply]
Looking through the OpenSim API documentation, it appears that the intention is for model designers to use the `ExternalLoads` and `ExternalForce` APIs to achieve this.

The way that those seem to be used is to create a separate document that specifies the external force parameters. E.g. this:

- https://github.com/opensim-org/opensim-core/blob/4.4/Applications/Forward/test/pendulum_external_load.xml

And then also separately provide an STO file containing the force vectors, e.g:

- https://github.com/opensim-org/opensim-core/blob/4.4/Applications/Forward/test/pendulum_external_load.sto

(e.g. yours could be at t = 0 to t=100000 with some constant force)

The `pendulum_external_load.xml` file is then provided to a tool, like the Forward Dynamic simulator tool in the official OpenSim GUI. My supervisor said that one way to handle this isn't to use an external forces file. That will work, but it's a fairly complicated way to handle it.

The recommended way, though, to add a constant force into the model is to add a `PrescribedForce`, as you tried, but you will need to assign `Constant` functions for each thing it requires (e.g. `Constant` 10 N in X).

OpenSim Creator doesn't have tooling UI for creating `OpenSim::Function`s visually yet, because it's more complicated to do that. You'll need to add it directly into  the osim file. I'll try my best to write a basic example file for you tomorrow but, if I can't do that, then the thing you'll need to search for is "OpenSim PrescribedForce example" or similar.

As a basic, concrete example of using a `PrescribedForce`, you can:

- Add a `PrescribedForce` into the model in OpenSim Creator
- Set its socket frame to the body you want to affect with the force
- Save the model somewhere
- Open the `.osim` file in a text editor
- Find the force you have just added
- Manually add functions that apply the force

E.g. here's an example of applying 10N along Y in the `.osim` file:

```xml
<PrescribedForce name="PrescribedForce">
    <!--Path to a Component that satisfies the Socket 'frame' of type PhysicalFrame (description: The frame onto which this force is applied.).-->
    <socket_frame>/bodyset/new_body</socket_frame>
    <FunctionSet name="forceFunctions">
        <objects>
            <Constant>
                <value>0</value>
            </Constant>
            <Constant>
                <value>10</value>
            </Constant>
            <Constant>
                <value>0</value>
            </Constant>
        </objects>
        <groups />
    </FunctionSet>
</PrescribedForce>
```

**Reply by UNNI-SANKAR** [reply]
Sir I have used the above code to create a gravity force due to a ball placed on hand in the arm26.osim file. i created a sphere of 5 kg in the hand placed using a free joint. I added the above code at the end of the main code of .osim file opened using the notepad. And after saving the above code in notepad i am getting the error message as shown below. can you explain why
![error](https://github.com/user-attachments/assets/220e89ee-c60c-4af5-bd16-d5ab8261b41d)
i also created the Prescribed force command in the opensim creator and after that i am getting the above error message

**Reply by adamkewley** [reply]
The `PrescribedForce` usually needs to be added to the `forceset` part of the osim file. It will cause XML errors if you paste elements at the end of the osim file.

I can add a `PrescribedForce` via OpenSim Creator to `arm26.osim` with no issues as described above
